[core] (cgroups 6/n) CgroupManager cleans up the entire cgroup hierarchy in reverse order when destroyed.

israbbani opened the pull request:
This PR stacks on #56246.

For more details about the resource isolation project see https://github.com/ray-project/ray/issues/54703.

This PR implements RAII cleanup in CgroupManager. If the CgroupManager is destroyed, it will undo all cgroup-related side-effects in reverse order. There are four kinds of side-effects
1. Creating Cgroups (reversed by deleting them)
2. Moving Processes between Cgroups (reversed by moving them back)
3. Enabling Controllers (reversed by disabling them)
4. Adding Resource Constraints (reversed by reverting to default resource constraints)

A few notes about clean up:
1. Clean up is currently best-effort meaning that if any part of it fails, it will log a warning and continue to attempt the rest of it.
2. Clean up _has_ to be done in order. For example, a cgroup cannot be deleted before its children are deleted. So cleanup proceeds from the leaf nodes up to the root note of the cgroup tree. 
3. The testing for cleanup needs to not only detect that clean up happened, but that it happened in the correct order. 

I've left comments in the PR outlining the most suspect aspects of the design. Happy to incorporate feedback.

Commit:
[core] (cgroups 1/n) Adding a sys/fs filesystem driver
to perform cgroup operations.

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
adding the copyright

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Adding a fallback for creating processes inside cgroups with fork/exec
instead of clone for older kernel headers < 5.7 (which is what we have
in CI)

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
adding a pause in the tests to see what's up with the container

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Update src/ray/common/cgroup2/cgroup_driver_interface.h

Co-authored-by: Edward Oakes <ed.nmi.oakes@gmail.com>
Signed-off-by: Ibrahim Rabbani <israbbani@gmail.com>

Commit:
Comments

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Merge branch 'irabbani/cgroups-1' of github.com:ray-project/ray into irabbani/cgroups-1

Commit:
Putting the cgroupv2 tests into a separate target

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
removing test sleep

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Removing a docstring

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
enabling CI tests

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
fixing absl imports

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
commenting local

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
doxygen formatting

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Merge branch 'master' into irabbani/cgroups-1

Commit:
removing integration tests

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
final cleanup

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
iwyu

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Merge branch 'master' into irabbani/cgroups-1

Commit:
we cpplintin!

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Update src/ray/common/cgroup2/sysfs_cgroup_driver.cc

Co-authored-by: Edward Oakes <ed.nmi.oakes@gmail.com>
Signed-off-by: Ibrahim Rabbani <israbbani@gmail.com>

Commit:
Apply suggestions from code review

Co-authored-by: Edward Oakes <ed.nmi.oakes@gmail.com>
Signed-off-by: Ibrahim Rabbani <israbbani@gmail.com>

Commit:
bug

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Merge branch 'irabbani/cgroups-1' of github.com:ray-project/ray into irabbani/cgroups-1

Commit:
[core] Integration tests for SysFsCgroupDriver.

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Cleaning up cgroup_test_utils and attempting to
fix CI.

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
broken

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
up

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
upup

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Merge branch 'master' into irabbani/cgroups-2

Signed-off-by: Ibrahim Rabbani <irabbani@anyscale.com>

Commit:
up

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Adding shell scripts to do cgroup setup/teardown
and enabling test for ci

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Merge branch 'master' into irabbani/cgroups-2

Commit:
Merged and fixed a few issues

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
fixing test target for CI

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
maybe this will trigger tests

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
runforever

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
up

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
up

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
cleaning up todos and docs

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
one more

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
adding separate target for unit tests now

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
typo

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
come unit and integration test targets

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
missing flag

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
plz work

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
one more

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
[core] Adding CgroupManager to create, modify, and delete the cgroup
hierarchy for ray.

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
disabling cgroup test

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Addressing feedback
- integration tests get their own directory
- using bazel supported platforms instead of a no_windows tag (which we
  can probably deprecate with this flag)

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
ci change

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Begrudingly using the random id generator from id.h

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
instructions for running locally

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
adding instructions to run locally

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Merge branch 'master' into irabbani/cgroups-2

Commit:
Merge branch 'irabbani/cgroups-2' into irabbani/cgroups-3

Signed-off-by: Ibrahim Rabbani <irabbani@anyscale.com>

Commit:
Cleaning up comments

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
fixing ci

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Merge branch 'irabbani/cgroups-2' of github.com:ray-project/ray into irabbani/cgroups-2

Commit:
Merge branch 'irabbani/cgroups-2' into irabbani/cgroups-3

Commit:
Removing the no_windows tags and replacing it with the bazel
target_compatible_with option. Cleaning up comments.

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
ci

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Merge branch 'irabbani/cgroups-3' of github.com:ray-project/ray into irabbani/cgroups-3

Commit:
Merge branch 'irabbani/cgroups-2' into irabbani/cgroups-3

Commit:
Merge branch 'master' into irabbani/cgroups-3

Signed-off-by: Ibrahim Rabbani <irabbani@anyscale.com>

Commit:
[core] cgroups (4/n) adding constraint bounds checking to the
CgroupManager. Removing Controller and Constraint checks from the
CgroupDriver.

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Merge branch 'master' into irabbani/cgroups-3

Commit:
Merge branch 'irabbani/cgroups-3' into irabbani/cgroups-4

Commit:
[core] (cgroups 5/n) Adding methods the following methods to
CgroupDriverInterface and SysFsCgroupDriver for cleanup along with unit
and integration tests

- DeleteCgroup
- DisableController

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Build with clang to find bugs locally!

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Merge branch 'irabbani/cgroups-4' of github.com:ray-project/ray into irabbani/cgroups-4

Commit:
Merge branch 'master' into irabbani/cgroups-3

Commit:
[core] (cgroups 6/n) CgroupManager cleans up the entire cgroup hierarchy
in reverse order when destroyed. Added unit tests.

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Adding a very long happy path test

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Merge branch 'irabbani/cgroups-3' into irabbani/cgroups-4

Commit:
Addressing feedback.

Signed-off-by: irabbani <irabbani@anyscale.com>

edoakes committed:
Merge branch 'master' into irabbani/cgroups-3

Commit:
Merge branch 'irabbani/cgroups-3' into irabbani/cgroups-4

Commit:
Merge branch 'irabbani/cgroups-4' into irabbani/cgroups-5

Commit:
Merge branch 'irabbani/cgroups-5' into irabbani/cgroups-6

israbbani commented on src/ray/common/cgroup2/fake_cgroup_driver.h:
```diff
@@ -39,31 +39,83 @@ struct FakeCgroup {
   }
 };
+
+struct FakeConstraint {
+  std::string cgroup_;
+  std::string name_;
+};
+
+struct FakeController {
+  std::string cgroup_;
+  std::string name_;
+};
+
+struct FakeMoveProcesses {
+  std::string from_;
+  std::string to_;
+};
```
I could fold these into a more compact single type, but I think it'll be less readable. Open to feedback.

Commit:
cleaning up and adding comments

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Merge branch 'irabbani/cgroups-6' of github.com:ray-project/ray into irabbani/cgroups-6

Commit:
Merge branch 'master' into irabbani/cgroups-5

Signed-off-by: Ibrahim Rabbani <irabbani@anyscale.com>

Commit:
Merge branch 'irabbani/cgroups-5' into irabbani/cgroups-6

Commit:
Merge branch 'master' into irabbani/cgroups-6

Signed-off-by: Ibrahim Rabbani <irabbani@anyscale.com>

Commit:
[core] (cgroups 7/n) cleaning up old cgroup integration code for raylet and core worker (#56285)

This PR stacks on #56260.

For more details about the resource isolation project see
https://github.com/ray-project/ray/issues/54703.

This PR removes previous code for integrating resource isolation with
the raylet and core worker. It's a prerequisite for integrating the
CgroupManager with the raylet.

This PR is cleanup only and there is no behavior change.

---------

Signed-off-by: irabbani <irabbani@anyscale.com>
Signed-off-by: Ibrahim Rabbani <irabbani@anyscale.com>

Commit:
Merge branch 'master' into irabbani/cgroups-6

Commit:
Merge branch 'master' into irabbani/cgroups-6

Commit:
Merge branch 'master' into irabbani/cgroups-6

ZacAttack reviewed (approved):
Synced offline, this looks good

Commit:
Merge branch 'master' into irabbani/cgroups-6

Commit:
[core] (cgroups 8/n) Wiring CgroupManager into the raylet. (#56297)

This PR stacks on #56285.

For more details about the resource isolation project see
https://github.com/ray-project/ray/issues/54703.

This PR integrates the CgroupManager with the raylet startup in
`main.cc`. It includes
- A cross-platform cgroup_manager bazel target that selectively compiles
dependencies for linux/non-linux platforms. I prefer this design because
it keeps targets small and compile times low and removes unnecessary
ifdefs from the code.
- Adds the following command-line args to the raylet binary
   - enable_resource_isolation
   - cgroup_path
   - system_reserved_cpu_weight
   - system_reserved_memory_bytes

I've left comments on the PR to highlight parts that could use some
feedback/discussion.

---------

Signed-off-by: irabbani <irabbani@anyscale.com>
Signed-off-by: Ibrahim Rabbani <irabbani@anyscale.com>

Commit:
Merge branch 'master' into irabbani/cgroups-6
